File: src/utils/mealPlan.js
/**
 * Meal plan helpers shared by the web Plan tab: meal/slot vocabularies,
 * Monday-start week math on local calendar days, and week ingredient merging.
 * Plan dates are local calendar days stored as `YYYY-MM-DD` strings.
 */

export const MEAL_TYPES = [
  { id: 'breakfast', label: 'Breakfast' },
  { id: 'lunch', label: 'Lunch' },
  { id: 'dinner', label: 'Dinner' },
];

export const ENTRY_KINDS = [
  { id: 'recipe', label: 'Recipe', icon: '📖' },
  { id: 'custom', label: 'Other meal', icon: '✏️' },
  { id: 'leftovers', label: 'Leftovers', icon: '🥡' },
  { id: 'eating_out', label: 'Eating out', icon: '🍽️' },
  { id: 'skip', label: 'Skip', icon: '⊘' },
];

const KIND_LABELS = Object.fromEntries(ENTRY_KINDS.map((k) => [k.id, k.label]));

/**
 * Text shown for a planned slot: the recipe name snapshot / custom title, or the kind's label.
 * @param {{kind: string, title?: string|null}} entry
 * @returns {string}
 */
export const entryDisplayTitle = (entry) => entry?.title || KIND_LABELS[entry?.kind] || '';

/**
 * Formats a Date as a local `YYYY-MM-DD` key.
 * @param {Date} date
 * @returns {string}
 */
export const toDateKey = (date) => {
  const y = date.getFullYear();
  const m = String(date.getMonth() + 1).padStart(2, '0');
  const d = String(date.getDate()).padStart(2, '0');
  return `${y}-${m}-${d}`;
};

/**
 * Parses a `YYYY-MM-DD` key as local midnight (not UTC, which `new Date(key)` would use).
 * @param {string} key
 * @returns {Date}
 */
export const fromDateKey = (key) => {
  const [y, m, d] = key.split('-').map(Number);
  return new Date(y, m - 1, d);
};

/**
 * The Monday starting the week that contains `date`, at local midnight.
 * @param {Date} date
 * @returns {Date}
 */
export const startOfWeek = (date) => {
  const day = new Date(date.getFullYear(), date.getMonth(), date.getDate());
  const offset = (day.getDay() + 6) % 7;
  day.setDate(day.getDate() - offset);
  return day;
};

/**
 * The seven days of the week starting at `weekStart`.
 * @param {Date} weekStart
 * @returns {Date[]}
 */
export const weekDays = (weekStart) =>
  Array.from({ length: 7 }, (_, i) => new Date(weekStart.getFullYear(), weekStart.getMonth(), weekStart.getDate() + i));

/**
 * Moves a week start by a number of weeks.
 * @param {Date} weekStart
 * @param {number} weeks
 * @returns {Date}
 */
export const shiftWeek = (weekStart, weeks) =>
  new Date(weekStart.getFullYear(), weekStart.getMonth(), weekStart.getDate() + weeks * 7);

/**
 * Key for a slot in an entries map.
 * @param {string} dateKey
 * @param {string} meal
 * @returns {string}
 */
export const slotKey = (dateKey, meal) => `${dateKey}|${meal}`;

/**
 * Merges ingredients across the week's planned recipes for the add-to-list flow.
 * Each distinct ingredient (case-insensitive name) appears once; its `amount` is the
 * number of planned recipe slots that use it, so a twice-planned recipe counts twice.
 * @param {string[]} plannedRecipeIds - Recipe id per planned slot, in week order (may repeat)
 * @param {Array<{recipeId: string, name: string, quantity?: string|null}>} ingredients
 * @returns {Array<{name: string, quantity: string|null, amount: number, unit: null}>}
 */
export const mergeWeekIngredients = (plannedRecipeIds, ingredients) => {
  const byRecipe = new Map();
  for (const ingredient of ingredients) {
    const list = byRecipe.get(ingredient.recipeId) ?? [];
    list.push(ingredient);
    byRecipe.set(ingredient.recipeId, list);
  }

  const merged = new Map();
  for (const recipeId of plannedRecipeIds) {
    for (const ingredient of byRecipe.get(recipeId) ?? []) {
      const key = ingredient.name.trim().toLowerCase();
      if (!key) continue;
      const existing = merged.get(key);
      if (existing) {
        existing.amount += 1;
      } else {
        merged.set(key, { name: ingredient.name.trim(), quantity: ingredient.quantity ?? null, amount: 1, unit: null });
      }
    }
  }
  return [...merged.values()];
};

File: ios-native/GatherLists/GatherLists/Models/MealPlan.swift
import Foundation

/// A meal of the day that can be planned.
enum MealType: String, Codable, CaseIterable, Identifiable, Hashable {
    case breakfast
    case lunch
    case dinner

    var id: String { rawValue }

    var label: String {
        switch self {
        case .breakfast: return "Breakfast"
        case .lunch: return "Lunch"
        case .dinner: return "Dinner"
        }
    }

    var systemImage: String {
        switch self {
        case .breakfast: return "sunrise"
        case .lunch: return "sun.max"
        case .dinner: return "moon.stars"
        }
    }
}

/// What occupies a planned meal slot.
enum MealEntryKind: String, Codable, CaseIterable, Hashable {
    case recipe
    case eatingOut = "eating_out"
    case leftovers
    case skip
    case custom

    var label: String {
        switch self {
        case .recipe: return "Recipe"
        case .eatingOut: return "Eating out"
        case .leftovers: return "Leftovers"
        case .skip: return "Skip"
        case .custom: return "Other meal"
        }
    }

    var systemImage: String {
        switch self {
        case .recipe: return "book"
        case .eatingOut: return "fork.knife"
        case .leftovers: return "takeoutbag.and.cup.and.straw"
        case .skip: return "minus.circle"
        case .custom: return "text.cursor"
        }
    }
}

/// A household meal plan, mapped to the `meal_plans` Supabase table.
struct MealPlan: Codable, Identifiable, Hashable {
    let id: UUID
    let ownerId: UUID
    var name: String
    var enabledMeals: [String]
    let createdAt: Date
    let updatedAt: Date

    /// Enabled meals in breakfast → dinner order, ignoring unknown values.
    var enabledMealTypes: [MealType] {
        MealType.allCases.filter { enabledMeals.contains($0.rawValue) }
    }

    enum CodingKeys: String, CodingKey {
        case id
        case ownerId = "owner_id"
        case name
        case enabledMeals = "enabled_meals"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }
}

/// A sharing record for a meal plan, mapped to the `meal_plan_shares` Supabase table.
struct MealPlanShare: Codable, Identifiable, Hashable {
    let id: UUID
    let mealPlanId: UUID
    let sharedWithEmail: String
    let sharedBy: UUID?
    let permission: String
    let addedAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case mealPlanId = "meal_plan_id"
        case sharedWithEmail = "shared_with_email"
        case sharedBy = "shared_by"
        case permission
        case addedAt = "added_at"
    }
}

/// One planned meal slot, mapped to the `meal_plan_entries` Supabase table.
/// `date` is a calendar day string (`yyyy-MM-dd`) with no time zone.
struct MealPlanEntry: Codable, Identifiable, Hashable {
    let id: UUID
    let mealPlanId: UUID
    var date: String
    var meal: MealType
    var kind: MealEntryKind
    var recipeId: UUID?
    var title: String?
    var note: String?
    var isLocked: Bool
    var source: String
    var cookedAt: Date?
    /// Why the planner picked this recipe; only set on `source == "suggested"` slots.
    var suggestionReason: String?
    let createdBy: UUID?
    let createdAt: Date
    var updatedAt: Date

    var isSuggested: Bool { source == "suggested" }

    /// Text shown for the slot: the recipe name snapshot, or the kind's label.
    var displayTitle: String {
        if let title, !title.isEmpty { return title }
        return kind.label
    }

    enum CodingKeys: String, CodingKey {
        case id
        case mealPlanId = "meal_plan_id"
        case date
        case meal
        case kind
        case recipeId = "recipe_id"
        case title
        case note
        case isLocked = "is_locked"
        case source
        case cookedAt = "cooked_at"
        case suggestionReason = "suggestion_reason"
        case createdBy = "created_by"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }
}

/// Calendar helpers for Monday-start meal plan weeks. Plan dates are local
/// calendar days, stored as `yyyy-MM-dd`.
enum MealPlanWeek {
    static var calendar: Calendar {
        var calendar = Calendar(identifier: .gregorian)
        calendar.firstWeekday = 2
        calendar.timeZone = .current
        return calendar
    }

    private static let formatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.calendar = Calendar(identifier: .gregorian)
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = .current
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter
    }()

    /// The Monday that starts the week containing `date`.
    static func startOfWeek(containing date: Date) -> Date {
        let cal = calendar
        let day = cal.startOfDay(for: date)
        let weekday = cal.component(.weekday, from: day)
        let offset = (weekday + 5) % 7
        return cal.date(byAdding: .day, value: -offset, to: day) ?? day
    }

    /// The seven days of the week starting at `weekStart`.
    static func days(from weekStart: Date) -> [Date] {
        (0..<7).compactMap { calendar.date(byAdding: .day, value: $0, to: weekStart) }
    }

    static func shift(_ weekStart: Date, byWeeks weeks: Int) -> Date {
        calendar.date(byAdding: .day, value: weeks * 7, to: weekStart) ?? weekStart
    }

    static func key(for date: Date) -> String {
        formatter.string(from: date)
    }

    static func date(fromKey key: String) -> Date? {
        formatter.date(from: key)
    }
}
